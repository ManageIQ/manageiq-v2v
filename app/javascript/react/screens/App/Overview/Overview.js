import React from 'react';
import PropTypes from 'prop-types';
import Equalizer from 'react-equalizer';
import { bindMethods, Button, EmptyState, Grid } from 'patternfly-react';
import componentRegistry from '../../../../components/componentRegistry';

class Overview extends React.Component {
  constructor(props) {
    super(props);

    bindMethods(this, ['getNodes']);

    this.mappingWizard = componentRegistry.markup(
      'MappingWizardContainer',
      props.store
    );
    this.planWizard = componentRegistry.markup(
      'PlanWizardContainer',
      props.store
    );
  }
  getNodes(equalizerComponent, equalizerElement) {
    return [this.node1, this.node2];
  }
  render() {
    const {
      showMappingWizardAction,
      showPlanWizardAction,
      mappingWizardVisible,
      planWizardVisible
    } = this.props;

    const overviewCards = (
      <div
        className="row cards-pf"
        ref={n => (this.mainContent = n)}
        style={{ overflow: 'auto', paddingBottom: 100, height: '100%' }}
      >
        <Grid.Col md={12}>
          <Grid.Row>
            <div className="spacer" />
            <Equalizer nodes={this.getNodes}>
              <Grid.Col xs={12} md={6}>
                <div className="card-pf">
                  <div className="card-pf-body">
                    <div className="blank-slate-pf" ref={n => (this.node1 = n)}>
                      <EmptyState.Icon />
                      <EmptyState.Title>
                        {__('Infrastructure Mappings')}
                      </EmptyState.Title>
                      <EmptyState.Info>
                        {__(
                          'Create mapping to later be used by a migration plan.'
                        )}
                      </EmptyState.Info>
                      <EmptyState.Action>
                        <Button
                          bsStyle="primary"
                          bsSize="large"
                          onClick={showMappingWizardAction}
                        >
                          {__('Create Infrastructure Mapping')}
                        </Button>
                      </EmptyState.Action>
                    </div>
                  </div>
                </div>
              </Grid.Col>
              <Grid.Col xs={12} md={6}>
                <div className="card-pf">
                  <div className="card-pf-body">
                    <div className="blank-slate-pf" ref={n => (this.node2 = n)}>
                      <EmptyState.Icon />
                      <EmptyState.Title>
                        {__('Migration Plans')}
                      </EmptyState.Title>
                      <EmptyState.Info>
                        {__('Create a migration plan to start migrating.')}
                      </EmptyState.Info>
                      <EmptyState.Action>
                        <Button
                          bsStyle="primary"
                          bsSize="large"
                          onClick={showPlanWizardAction}
                        >
                          {__('Create Migration Plan')}
                        </Button>
                      </EmptyState.Action>
                    </div>
                  </div>
                </div>
              </Grid.Col>
            </Equalizer>
          </Grid.Row>

          <Grid.Row>
            <Grid.Col xs={12}>
              <div className="card-pf">
                <div className="card-pf-heading">
                  <h2>0 Migrations in Progress</h2>
                </div>
                <div className="card-pf-body">
                  <p>content</p>
                </div>
              </div>
            </Grid.Col>
          </Grid.Row>

          <Grid.Row>
            <Grid.Col xs={12}>
              <div className="card-pf">
                <div className="card-pf-heading">
                  <h2>0 Completed Migrations</h2>
                </div>
                <div className="card-pf-body">
                  <p>content</p>
                </div>
              </div>
              <div className="spacer" />
            </Grid.Col>
          </Grid.Row>
        </Grid.Col>
      </div>
    );

    return (
      <React.Fragment>
        {overviewCards}
        {mappingWizardVisible && this.mappingWizard}
        {planWizardVisible && this.planWizard}
      </React.Fragment>
    );
  }
}
Overview.propTypes = {
  store: PropTypes.object,
  showMappingWizardAction: PropTypes.func,
  showPlanWizardAction: PropTypes.func,
  mappingWizardVisible: PropTypes.bool,
  planWizardVisible: PropTypes.bool
};
export default Overview;
