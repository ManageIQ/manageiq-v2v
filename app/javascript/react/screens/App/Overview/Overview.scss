@import './screens/MappingWizard/MappingWizard.scss';
@import './screens/PlanWizard/components/DropdownButtonSelect/DropdownButtonSelect.scss';
@import './screens/PlanWizard/components/PlanWizardResultsStep/PlanWizardResultsStep.scss';
@import 'components/Cards/MigrationsInProgressCard/MigrationsInProgress.scss';
@import 'components/Cards/MigrationsCompletedCard/MigrationsCompleted.scss';
@import 'components/ModalWizard/ModalWizard.scss';

.card-pf-body .blank-slate-pf {
  background-color: #fff;
}

@media (min-width: 768px) {
  .card-pf-body .blank-slate-pf {
    padding: 20px 20px;
  }
}

.csv-upload-dropzone {
  border: 3px solid #fff;
}

.csv-upload-dropzone.active {
  border: 3px dashed #9b9b9b; // Matches upload icon color
}

.cards-pf .blank-slate-pf {
  border: none;

  .blank-slate-pf-icon {
    color: #d1d1d1;
  }
}
