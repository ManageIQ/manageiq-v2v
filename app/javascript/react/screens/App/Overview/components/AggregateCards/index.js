import MigrationsNotStarted from './MigrationsNotStarted/MigrationsNotStarted';
import MigrationsInProgress from './MigrationsInProgress/MigrationsInProgress';
import MigrationsComplete from './MigrationsComplete/MigrationsComplete';
import InfrastructureMappings from './InfrastructureMappings/InfrastructureMappings';

export {
  MigrationsNotStarted,
  MigrationsInProgress,
  MigrationsComplete,
  InfrastructureMappings
};
