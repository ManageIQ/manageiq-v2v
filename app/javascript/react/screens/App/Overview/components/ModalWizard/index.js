import ModalWizard from './ModalWizard';
import ModalWizardBody from './ModalWizardBody';

ModalWizard.Body = ModalWizardBody;

export default ModalWizard;
