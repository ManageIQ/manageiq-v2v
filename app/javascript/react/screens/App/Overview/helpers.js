export const planRequestWithPlanId = (
  transformationPlanRequest,
  transformationPlans
) => {
  const matchingPlan = transformationPlans.find(transformationPlan =>
    transformationPlan.miq_requests.some(
      request => request.id === transformationPlanRequest.id
    )
  );

  return matchingPlan
    ? { ...transformationPlanRequest, plan_id: matchingPlan.id }
    : transformationPlanRequest;
};

export const planRequestsWithPlanIds = (
  transformationPlanRequests,
  transformationPlans
) =>
  transformationPlanRequests.map(transformationPlanRequest =>
    planRequestWithPlanId(transformationPlanRequest, transformationPlans)
  );
