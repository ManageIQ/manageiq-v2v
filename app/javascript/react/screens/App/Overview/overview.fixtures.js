import Immutable from 'seamless-immutable';

export const initialState = Immutable({
  mappingWizardVisible: false,
  hideMappingWizard: false,
  planWizardVisible: false,
  hidePlanWizard: false
});
