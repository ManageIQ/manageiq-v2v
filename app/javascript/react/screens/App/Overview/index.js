import { connect } from 'react-redux';
import Overview from './Overview';
import * as OverviewActions from './OverviewActions';

import reducer from './OverviewReducer';

export const reducers = { overview: reducer, form: {} };

const mapStateToProps = ({ overview }, ownProps) => ({
  ...overview,
  ...ownProps.data
});

const mergeProps = (stateProps, dispatchProps, ownProps) =>
  Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(mapStateToProps, OverviewActions, mergeProps)(Overview);
