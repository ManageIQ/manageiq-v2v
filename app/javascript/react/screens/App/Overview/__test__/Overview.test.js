import React from 'react';
import { shallow } from 'enzyme';
import Overview from '../Overview';
import { coreComponents } from '../../../../../components';
import componentRegistry from '../../../../../components/componentRegistry';
import { transformationMappings } from '../overview.fixtures';

jest.mock('../../../../../components/componentRegistry');
jest.useFakeTimers();

componentRegistry.registerMultiple(coreComponents);

describe('Overview component', () => {
  const baseProps = {
    store: {},
    mappingWizardVisible: false,
    planWizardVisible: false,
    isFetchingTransformationMappings: false,
    isRejectedTransformationMappings: false,
    isFetchingTransformationPlanRequests: false,
    isRejectedTransformationPlanRequests: false,
    errorTransformationPlanRequests: null,
    transformationMappings: []
  };
  let showMappingWizardAction;
  let showPlanWizardAction;
  let fetchTransformationMappingsAction;
  let fetchTransformationPlanRequestsAction;
  let fetchTransformationPlansAction;
  beforeEach(() => {
    showMappingWizardAction = jest.fn();
    showPlanWizardAction = jest.fn();
    fetchTransformationMappingsAction = jest.fn();
    fetchTransformationPlanRequestsAction = jest.fn();
    fetchTransformationPlansAction = jest.fn();
  });

  describe('overview sections', () => {
    test('does not render Migrations if there are no transformation mappings', () => {
      const wrapper = shallow(
        <Overview
          {...baseProps}
          showMappingWizardAction={showMappingWizardAction}
          showPlanWizardAction={showPlanWizardAction}
          fetchTransformationMappingsAction={fetchTransformationMappingsAction}
          fetchTransformationPlanRequestsAction={
            fetchTransformationPlanRequestsAction
          }
          fetchTransformationPlansAction={fetchTransformationPlansAction}
        />
      );

      expect(wrapper.find('Migrations').exists()).toBe(false);
    });

    test('renders Migrations if there are transformation mappings', () => {
      const wrapper = shallow(
        <Overview
          {...baseProps}
          transformationMappings={transformationMappings}
          showMappingWizardAction={showMappingWizardAction}
          showPlanWizardAction={showPlanWizardAction}
          fetchTransformationMappingsAction={fetchTransformationMappingsAction}
          fetchTransformationPlanRequestsAction={
            fetchTransformationPlanRequestsAction
          }
          fetchTransformationPlansAction={fetchTransformationPlansAction}
        />
      );

      expect(wrapper.find('Migrations').exists()).toBe(true);
    });
  });

  describe('polling', () => {
    let wrapper; // eslint-disable-line no-unused-vars
    beforeEach(() => {
      wrapper = shallow(
        <Overview
          {...baseProps}
          showMappingWizardAction={showMappingWizardAction}
          showPlanWizardAction={showPlanWizardAction}
          fetchTransformationMappingsAction={fetchTransformationMappingsAction}
          fetchTransformationPlanRequestsAction={
            fetchTransformationPlanRequestsAction
          }
          fetchTransformationPlansAction={fetchTransformationPlansAction}
        />
      );
    });
    test('starts when the component is mounted', () => {
      expect(fetchTransformationPlanRequestsAction).toHaveBeenCalledTimes(1);
    });

    test('fetches transformation plan requests every 15 seconds', () => {
      jest.advanceTimersByTime(15000);
      expect(fetchTransformationPlanRequestsAction).toHaveBeenCalledTimes(2);
    });

    // TODO: Come back to these once the UI is closer to final form
    test.skip('stops if the mapping wizard is visible', () => {});

    test.skip('stops if the plan wizard is visible', () => {});
  });
});
