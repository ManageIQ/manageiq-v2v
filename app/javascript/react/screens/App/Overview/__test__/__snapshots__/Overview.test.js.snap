// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview component renders the overview 1`] = `
<React.Fragment>
  <div
    className="row cards-pf"
    style={
      Object {
        "height": "100%",
        "overflow": "auto",
        "paddingBottom": 100,
      }
    }
  >
    <Col
      bsClass="col"
      componentClass="div"
      md={12}
    >
      <Row
        bsClass="row"
        componentClass="div"
      >
        <div
          className="spacer"
        />
        <Equalizer
          byRow={true}
          enabled={[Function]}
          nodes={[Function]}
          property="height"
        >
          <Col
            bsClass="col"
            componentClass="div"
            md={6}
            xs={12}
          >
            <div
              className="card-pf"
            >
              <div
                className="card-pf-body"
              >
                <div
                  className="blank-slate-pf"
                >
                  <EmptyStateIcon
                    name="add-circle-o"
                    type="pf"
                  />
                  <EmptyStateTitle>
                    Infrastructure Mappings
                  </EmptyStateTitle>
                  <EmptyStateInfo>
                    Create mapping to later be used by a migration plan.
                  </EmptyStateInfo>
                  <EmptyStateAction>
                    <Button
                      bsSize="large"
                      bsStyle="primary"
                      onClick={[MockFunction]}
                    >
                      Create Infrastructure Mapping
                    </Button>
                  </EmptyStateAction>
                </div>
              </div>
            </div>
          </Col>
          <Col
            bsClass="col"
            componentClass="div"
            md={6}
            xs={12}
          >
            <div
              className="card-pf"
            >
              <div
                className="card-pf-body"
              >
                <div
                  className="blank-slate-pf"
                >
                  <EmptyStateIcon
                    name="add-circle-o"
                    type="pf"
                  />
                  <EmptyStateTitle>
                    Migration Plans
                  </EmptyStateTitle>
                  <EmptyStateInfo>
                    Create a migration plan to start migrating.
                  </EmptyStateInfo>
                  <EmptyStateAction>
                    <Button
                      bsSize="large"
                      bsStyle="primary"
                      onClick={[MockFunction]}
                    >
                      Create Migration Plan
                    </Button>
                  </EmptyStateAction>
                </div>
              </div>
            </div>
          </Col>
        </Equalizer>
      </Row>
      <Row
        bsClass="row"
        componentClass="div"
      >
        <Col
          bsClass="col"
          componentClass="div"
          xs={12}
        >
          <div
            className="card-pf"
          >
            <div
              className="card-pf-heading"
            >
              <h2>
                0 Migrations in Progress
              </h2>
            </div>
            <div
              className="card-pf-body"
            >
              <p>
                content
              </p>
            </div>
          </div>
        </Col>
      </Row>
      <Row
        bsClass="row"
        componentClass="div"
      >
        <Col
          bsClass="col"
          componentClass="div"
          xs={12}
        >
          <div
            className="card-pf"
          >
            <div
              className="card-pf-heading"
            >
              <h2>
                0 Completed Migrations
              </h2>
            </div>
            <div
              className="card-pf-body"
            >
              <p>
                content
              </p>
            </div>
          </div>
          <div
            className="spacer"
          />
        </Col>
      </Row>
    </Col>
  </div>
</React.Fragment>
`;
